### DISCUSSION TITLE
Generate Normal Maps

### DISCUSSION THREAD

**Discussion by thraidh** [hlorus/CAD_Sketcher#71] [Feature Requests]
**Motivation / Existing Problem**
When doing hard surface modeling for game assets today, one of the most annoying problems for me is, when the shading breaks because of some unfortunate mesh topology.
This is especially annoying, since I may have started with a cylinder, but that is immediately broken into faces and therefore we lose the information about the normals. This only gets worse the more the cylinder is modified.

**Description**
Since the sketch is completely defined by math, we know the exact normal at each point. The request is to have a feature that populates normal maps with that information, so the user ends up with perfect normals at each location regardless of the mesh topology.

**Alternative Solutions**
An alternative would be to generate a high poly mesh, generate normals based on that mesh, generate a low poly mesh and bake the high-poly normals onto the low-poly mesh. That is cumbersome and we still lose the perfect information that we had.

**Comment by Intuos5** [comment]
So, if I understand this one correctly, the point is that the addon should create custom normals? How does this then interact with Blender's shade smooth/ flat and autosmooth?

**Reply by thraidh** [reply]
The idea is, that you need to UV unwrap and then those maps can be populated with exact normal information obtained from the original geometric forms.
Autosmooth does something similar, but afaik only uses the positions of connected vertices to calculate the normal for a vertex or face and then interpolates. It may be smarter than that, but the fact remains, that it can only use the vertex positions to do that.

But your point is adding a good idea. Since Blender can generate normals even without UV unwrapping, it may be possible to replace the smooth shader with something more precise for objects built on sketches.

**Comment by Intuos5** [comment]
Using UVs for this seems to be rather hacky. There's also the weighted normals modifier to help out with normals and shading issues. I think it would be better to create the normals using that and some form of smooth shading by angle (and mark smooth by tangency constrant, mark sharp by coincident) something like that.

**Reply by thraidh** [reply]
My goal is to create low-poly game assets with perfect normals with the least amount of effort, so I eventually need normal maps anyway, therefore this was my first approach.
But I don't know enough about Blender internals to make a really qualified suggestion on how to get to that goal.

I know that shading problems on hard-surface models today are usually fixed by adding/changing topology (like collapsing faces that are too small). Anyway, I think this is just to help the smooth shader to guess the intended normals. The weighted normals modifier still only uses vertex positions and seams to guess the intended normals, but gives more control over the guessing algorithm.

My reasoning now is, that the shader should not guess normals at all, because we can calculate the precise normals at every point.

Of course that only works as long as we are using defined mathematical operations to describe what we did. As soon as "destructive" direct modeling is used (like moving a vertex manually), we can only rely in vertex positions (and seams) again and would have to fall back to the imprecise guessing.

Maybe a less complicated solution would be to develop a modifier like the weighted normals modifier, that would not generate normals based on seams and vertex position, but use the sketch and other modifiers. This would still cause the normals only to be perfect at the  vertices and interpolated in between. This may or may not be close enough. I fear, though, that normal errors would become larger the lower the polycount. But it would probably still be worth trying.
